### PR TITLE
Added Example for Object Detection using Trained Tensorflow Models (SavedModel Format) 

### PR DESCRIPTION
**Pull request by uniquetrij**
## Description ##
Added a simple example to use DJL to perform object detection using models trained in Tensorflow and exported to SavedModel format. The example includes a custom translator to correctly load models from the [TensorFlow 2 Detection Model Zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md) 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
    - For new examples, [README.md](https://github.com/uniquetrij/djl/blob/master/examples/docs/object_detection_with_tensorflow_saved_model.md) is added to explain what the example does.
- [x] To the my best knowledge, [examples](https://github.com/awslabs/djl/tree/master/examples) and [jupyter notebooks](https://github.com/awslabs/djl/tree/master/jupyter) are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by frankfliu** [comment]
@uniquetrij 
Thank you for your contribution. There is a challenge to merge this PR. I will make some changes on top of your PR tomorrow.

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.examples.inference;
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:13]: FYI: We use Google AOSP coding format. You can format java code with following command:
```
./gradlew formatJava
```

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+    private ObjectDetectionWithTensorflowSavedModel() {}
+
+    public static void main(String[] args) throws IOException, ModelException, TranslateException {
+        DetectedObjects detection = ObjectDetectionWithTensorflowSavedModel.predict();
+        logger.info("{}", detection);
+    }
+
+    public static DetectedObjects predict() throws IOException, ModelException, TranslateException {
+        if (!"TensorFlow".equals(Engine.getInstance().getEngineName())) {
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:81]: FYI: Examples project by default doesn't include tensorflow engine dependency. User has to use the following command to use TF engine:
```
./gradlew :examples:run -Dai.djl.default_engine=TensorFlow
```

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        Image img = ImageFactory.getInstance().fromFile(imageFile);
+
+        String modelUrl =
+                "http://download.tensorflow.org/models/object_detection/tf2/20200711/ssd_mobilenet_v2_320x320_coco17_tpu-8.tar.gz";
+
+        Criteria<Image, DetectedObjects> criteria =
+                Criteria.builder()
+                        .optApplication(Application.CV.OBJECT_DETECTION)
+                        .setTypes(Image.class, DetectedObjects.class)
+                        .optModelUrls(modelUrl)
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:95]: FYI: optModelUrls() is preferred way to specify url, the system properties will impact the jvm globally and not thread-safe if you change it at runtime.

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        Image newImage = img.duplicate(Image.Type.TYPE_INT_ARGB);
+        newImage.drawBoundingBoxes(detection);
+
+        Path imagePath = outputDir.resolve("detected-tensorflow-model-dog_bike_car.png");
+        // OpenJDK can't save jpg with alpha channel
+        newImage.save(Files.newOutputStream(imagePath), "png");
+        logger.info("Detected objects image has been saved in: {}", imagePath);
+    }
+
+    static Map<Integer, String> loadSynset() throws IOException {
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:125]: FYI: DJL defines synset as List is just provide a convenient way to create ImageTranslator, you can load your label mapping in whatever way you want in your custom Translator

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        URL synsetUrl =
+                new URL(
+                        "https://raw.githubusercontent.com/tensorflow/models/master/research/object_detection/data/mscoco_label_map.pbtxt");
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:128]: FYI: the raw pbtxt file is a well-formatted file, it's much easier than passing the HTML file.

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    private static final class Item {
+        int id;
+
+        @SerializedName("display_name")
+        String displayName;
+    }
+
+    private static final class MyTranslator implements Translator<Image, DetectedObjects> {
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:154]: You don't have to extends from ObjecDetectionTranslator

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            array = NDImageUtils.resize(array, 224);
+            // tf object-detection models expect 8 bit unsigned integer tensor
+            array = array.toType(DataType.UINT8, true);
+            array = array.expandDims(0); // tf object-detection models expect a 4 dimensional input
+            return new NDList(array);
+        }
+
+        @Override
+        public void prepare(NDManager manager, Model model) throws IOException {
+            if (classes == null) {
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:180]: FYI: parepare() is the place to do one time initialization of the Translator

File: examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+                classes = loadSynset();
+                maxClassId = Collections.max(classes.keySet());
+            }
+        }
+
+        @Override
+        public DetectedObjects processOutput(TranslatorContext ctx, NDList list) {
+            // output of tf object-detection models is a list of tensors, hence NDList in djl
+            // output NDArray order in the list are not guaranteed
+            int length = 0; // number of boxes in the output
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/ObjectDetectionWithTensorflowSavedModel.java:190]: FYI: The output order is random especially during debug. Have to do some magic here.

**Comment by uniquetrij** [comment]
Thanks a lot for your help Frank. I'd like to continue to add more like
this as I continue to experiment more with DJL in the upcoming weeks.

Regards,
Trijeet Modak

**Comment by uniquetrij** [comment]
Updated the example with tensorflow.

**Review by roywei — APPROVED** [review]
Thanks for the contribution! We need this example, so we will merge this PR and make modifications afterwards if necessary.

**Pull request merged**